porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go

```diff
@@ -0,0 +1,436 @@
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/yaml"
+)
+
+type Options struct {
+	// BindFunction is the image name for the "bind" function
+	BindFunction string
+}
+
+func (o *Options) InitDefaults() {
+	o.BindFunction = "bind"
+}
+
+func (o *Options) BindFlags(prefix string, flags *flag.FlagSet) {
+	flag.StringVar(&o.BindFunction, "bindFunction", o.BindFunction, "image name for the bind function")
+}
+
+// KlippyReconciler reconciles Klippy objects
+type KlippyReconciler struct {
+	Options
+
+	client client.Client
+
+	dynamicClient dynamic.Interface
+	restMapper    meta.RESTMapper
+}
+
+// Reconcile implements the main kubernetes reconciliation loop.
+func (r *KlippyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	log.Info("reconciling object", "id", req.NamespacedName)
+
+	var parent api.PackageRevision
+	if err := r.client.Get(ctx, req.NamespacedName, &parent); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// TODO: Can we filter the watch?
+	if !parent.Status.Deployment {
+		log.V(2).Info("ignoring package not in deployments repository", "package", parent.Spec.PackageName)
+		return ctrl.Result{}, nil
+	}
+
+	var parentResources api.PackageRevisionResources
+	if err := r.client.Get(ctx, req.NamespacedName, &parentResources); err != nil {
+		// Not found here is unexpected
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcile(ctx, &parent, &parentResources); err != nil {
+		// TODO: raise event?
+		log.Error(err, "error reconciling", "id", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func isWhitespace(s string) bool {
+	for _, r := range s {
+		if !unicode.IsSpace(r) {
+			return false
+		}
+	}
+	return true
+}
+
+func (r *KlippyReconciler) reconcile(ctx context.Context, parent *api.PackageRevision, parentResources *api.PackageRevisionResources) error {
+	log := log.FromContext(ctx)
+
+	var blueprintPackageRevisions api.PackageRevisionList
+	if err := r.client.List(ctx, &blueprintPackageRevisions); err != nil {
+		return fmt.Errorf("error listing blueprints: %w", err)
+	}
+
+	// TODO: We should be able to cache this
+	blueprints, err := r.parseBlueprints(ctx, &blueprintPackageRevisions)
+	if err != nil {
+		return err
+	}
+
+	parentObjects, err := parseObjects(ctx, parentResources)
+	if err != nil {
+		return err
+	}
+
+	var allProposals []*api.PackageRevision
+	for _, blueprint := range blueprints {
+		proposals, err := r.buildProposals(ctx, parent, parentObjects, blueprint)
```

Comment:
seems like this is the part we would like to externalize (delegate to proposal makers).